### PR TITLE
Check the authentication level of user and resources

### PR DESCRIPTION
**Pull request by urshala**
* Add mid strength authentication level to the choice of authentication for resource.
* Check that user's logged in with correct authentication level can reserve correct resource.
   E.g. Google is weak authentication. If a user logged in via Google tries to reserve a resource
   whose authentication value is `strong`, then it should throw an error. However, this user should
   be able to reserve the resource whose authentication is `weak` or `none`.

### CODE REVIEW

File: locale/fi/LC_MESSAGES/django.po [hunk 1]
```diff
@@ -949,6 +949,9 @@ msgstr "Ei mitään"
 msgid "Weak"
 msgstr "Heikko"
 
+msgid "Mid"
+msgstr "keskivahvo"
```
Comment (jopesy) [inline locale/fi/LC_MESSAGES/django.po]: "Keskivahva"

File: locale/fi/LC_MESSAGES/django.po [hunk 2]
```diff
@@ -526,6 +526,12 @@ msgstr "Virheellinen pk \"{pk_value}\" - kohdetta ei ole olemassa."
 msgid "Invalid value in filter %(filter)s"
 msgstr "Virheellinen arvo filtterissä %(filter)s"
 
+msgid "You need to login with strong authentication to reserve this resource."
+msgstr ""
```
Comment (jopesy) [inline locale/fi/LC_MESSAGES/django.po]: "Sinun täytyy kirjautua vahvalla tunnistautumistavalla varataksesi tämän resurssin"

File: locale/fi/LC_MESSAGES/django.po [hunk 3]
```diff
@@ -526,6 +526,12 @@ msgstr "Virheellinen pk \"{pk_value}\" - kohdetta ei ole olemassa."
...
+
+msgid "You need to login with mid authentication to reserve this resource."
+msgstr ""
```
Comment (jopesy) [inline locale/fi/LC_MESSAGES/django.po]: "Sinun täytyy kirjautua keskivahvalla tunnistautumistavalla varataksesi tämän resurssin"

File: locale/sv/LC_MESSAGES/django.po [hunk 4]
```diff
@@ -522,6 +522,12 @@ msgstr "Felaktig pk \"{pk_value}\" - objektet existerar inte."
 msgid "Invalid value in filter %(filter)s"
 msgstr "Felaktigt värde i filter %(filter)s"
 
+msgid "You need to login with strong authentication to reserve this resource."
+msgstr ""
```
Comment (jopesy) [inline locale/sv/LC_MESSAGES/django.po]: "Du måste logga in med en stark autentiseringsmetod för att reservera den här resursen"

File: locale/sv/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -522,6 +522,12 @@ msgstr "Felaktig pk \"{pk_value}\" - objektet existerar inte."
...
+
+msgid "You need to login with mid authentication to reserve this resource."
+msgstr ""
```
Comment (jopesy) [inline locale/sv/LC_MESSAGES/django.po]: "Du måste logga in med en medelstark autentiseringsmetod för att reservera den här resursen"

File: locale/fi/LC_MESSAGES/django.po (hunk 2 above)
Comment (jopesy) [reply locale/fi/LC_MESSAGES/django.po]: Also these messages are not very user-friendly. Maybe the login methods they should use should be included in the message? Just an idea for improvement, could be a follow-up ticket to this one.

File: resources/api/reservation.py [hunk 6]
```diff
@@ -508,6 +509,66 @@ def has_object_permission(self, request, view, obj):
...
+
+    User's auth level       Reserveable resource with auth level
+      Strong          ->       strong, mid, weak, none
+      Mid             ->       mid, weak, none
+      Weak            ->       weak, none
+    """
+    message = ''
+    # TODO: Needs discussing with client which login method belongs to which level.
+    STRONG_AUTHENTICATION = ('suomifi')
+    MID_AUTHENTICATION = ('axiell_aurora')
```
Comment (tuomas777) [inline resources/api/reservation.py]: These should probably be tuples?
```suggestion
    STRONG_AUTHENTICATION = ('suomifi',)
    MID_AUTHENTICATION = ('axiell_aurora',)
```

**Review by tuomas777 — COMMENTED** [review]
Especially as this is about permission checks, it would be good to have tests as well.

File: resources/api/reservation.py [hunk 7]
```diff
@@ -508,6 +509,66 @@ def has_object_permission(self, request, view, obj):
...
+        ):
+            if resource_authentication == 'mid':
+                self.message = _('You need to login with mid authentication to reserve '
+                                 'this resource.')
+            else:
+                self.message = _('You need to login with strong authentication to '
+                                 'reserve this resource.')
+            return False
+
+        return True
```
Comment (jopesy) [inline resources/api/reservation.py]: I think a good practise with permission checks is to default to `False` and only deviate from that if a specific condition applied. That makes sure you're not giving permission in some edge case that you just hadn't thought of and the if statements don't catch.

```suggestion
    def _can_reserve_resource_with_current_login(self, request, resource):
            resource_authentication = resource.authentication
            user_authentication = get_user_auth_backend(request)
    
            if resource_authentication == 'none':
                return True
    
            if user_authentication in self.STRONG_AUTHENTICATION:
                return True
    
            if user_authentication in self.MID_AUTHENTICATION:
                if resource_authentication in ["mid", "weak"]:
                    return True
                else:
                    self.message = _('You need to login with strong authentication to reserve '
                                    'this resource.')
    
            if user_authentication in self.WEAK_AUTHENTICATION:
                if resource_authentication == "weak":
                    return True
                elif resource_authentication == 'mid':
                    self.message = _('You need to login with mid authentication to reserve '
                                     'this resource.')
                else:
                    self.message = _('You need to login with strong authentication to '
                                     'reserve this resource.')
    
            return False
```

Comment (jopesy) [reply resources/api/reservation.py]: Eh, don't know why Github doesn't like to show the diff of the suggestions in a nicer way.

**Comment by jopesy** [comment]
Left some comments, and adding a test for the permissions checks would be good as Tuomas suggested. Otherwise this looks good to me.

**Comment by urshala** [comment]
@jopesy @tuomas777 
Thank you for the review. 
I have made the suggested changes and added test now. Could you pls have one more look at it.

File: locale/fi/LC_MESSAGES/django.po (hunk 2 above)
Comment (urshala) [reply locale/fi/LC_MESSAGES/django.po]: I didn't change it here. I will update it in follow up ticket.

**Review by jopesy — APPROVED** [review]
Looks nice. Good work Deepak!

**Pull request merged**